apiserver/facades/client/charms/client.go

```diff
@@ -380,6 +412,77 @@ func (a *API) addCharmWithAuthorization(args params.AddCharmWithAuth) (params.Ch
 	}, nil
 }
 
+func (a *API) queueAsyncCharmDownload(args params.AddCharmWithAuth) (corecharm.Origin, error) {
+	charmURL, err := charm.ParseURL(args.URL)
+	if err != nil {
+		return corecharm.Origin{}, err
+	}
+
+	// Check if a charm doc already exists for this charm URL. If so, the
+	// charm has already been queued for download so this is a no-op. We
+	// still need to return back the recorded origin from where the charm
+	// is to be downloaded.
+	if _, err := a.backendState.Charm(charmURL); err == nil {
+		// Try to lookup the origin associated with the application.
+		// Due to the way that charm deploys are implemented by the
+		// client, if the actual deploy call fails, the resolved origin
+		// returned by this method will not be recorded against the
+		// application (as the application doc will not be created).
+		//
+		// If we are unable to figure out the origin, we still need
+		// to execute the remainder of this method
+		resolvedOrigin, err := a.backendState.UploadedCharmOrigin(charmURL)
+		if err == nil {
+			return resolvedOrigin, nil
+		} else if !errors.IsNotFound(err) {
+			return corecharm.Origin{}, errors.Trace(err)
+		}
+	}
+
+	// Fetch the charm metadata and add charm entry pending to be downloaded.
+	requestedOrigin := convertParamsOrigin(args.Origin)
+	repo, err := a.getCharmRepository(requestedOrigin.Source)
+	if err != nil {
+		return corecharm.Origin{}, errors.Trace(err)
+	}
+
+	// TODO(achilleasa):
+	// At this stage we are only interested in the charm metadata and lxd
+	// profile. However, the repo does not (yet) support metadata-only
+	// lookups so we will use the download API and extract the metadata
+	// we need.
```

Comment:
You can...

```
curl -XPOST -s https://api.charmhub.io/v2/charms/refresh -H 'Content-type: application/json' -d '{"context": [], "actions": [{"name": "juju-controller", "base": {"name": "ubuntu", "channel": "16.04", "architecture": "amd64"}, "channel":"beta", "action": "download", "instance-key": "a-test"}], "fields":["metadata-yaml"]}' | jq
```

Reply:
Indeed, both store APIs support this functionality (although only charmhub supports batch lookups IIRC). The next step forward is to implement a batch metadata lookup method in the repository interface which the new deploy code will use (group requirements by store and make batch lookup requests).

Since this facade will eventually be deprecated and superseded by the new deploy code, I didn't want to spend time (in this PR) to get that working so I opted for this hacky shortcut which has the same effect (with the obvious caveat that we fetch and drop the blob but this is all behind a feature flag so it's fine for now).